2996_reflection/reflection.md

```diff
@@ -271,4 +183,202 @@ std::string name1 = name_of(^S);             // Also okay.
+```
+:::
+
+
+## Selecting Members
+
+Our second example enables selecting a member "by number" for a specific type.  It also shows the use of a metafunction dealing with diagnostics:
+
+:::bq
+```c++
+struct S { unsigned i:2, j:6; };
+
+consteval auto member_number(int n) {
+  if (n == 0) return ^S::i;
+  else if (n == 1) return ^S::j;
+  else return std::meta::invalid_reflection("Only field numbers 0 and 1 permitted");
+}
+
+int main() {
+  S s{0, 0};
+  s.[:member_number(1):] = 42;  // Same as: s.j = 42;
+  s.[:member_number(5):] = 0;   // Error (likely with "Only field numbers 0 and 1 permitted" in text).
+}
+```
+:::
+
+This example also illustrates that bit fields are not beyond the reach of this proposal.
+
+
+## Fast Generation of Integer Sequence
+
+:::bq
+```c++
+#include <utility>
+#include <vector>
+
+template<typename T>
+consteval info make_integer_seq_refl(T N) {
+  std::vector args{^T};
+  for (T k = 0; k<N; ++k)  args.push_back(std::meta::reflect_value(k));
+  return substitute(^std::integer_sequence, args);
+}
+
+template<typename T>
+  using make_integer_sequence<typename T, T N> = [:make_integer_seq_refl<T>(N):];
+```
+:::
+
+
+## Enum to String
+
+One of the most commonly requested facilities is to convert an enum value to a string (this example relies on expansion statements [@P1306R1]):
+
+::: bq
+```c++
+template <typename E>
+  requires std::is_enum_v<E>
+constexpr std::string enum_to_string(E value) {
+  template for (constexpr auto e : std::meta::members_of(^E)) {
+    if (value == [:e:]) {
+      return std::string(std::meta::name_of(e));
+    }
+  }
+
+  return "<unnamed>";
+}
+
+enum Color { red, green, blue };
+static_assert(enum_to_string(Color::red) == "red");
+static_assert(enum_to_string(Color(42)) == "<unnamed>");
+```
+:::
+
+We can also do the reverse in pretty much the same way:
+
+::: bq
+```c++
+template <typename E>
+  requires std::is_enum_v<E>
+constexpr std::optional<E> string_to_enum(std::string_view name) {
+  template for (constexpr auto e : std::meta::members_of(^E)) {
+    if (name == std::meta::name_of(e)) {
+      return [:e:];
+    }
+  }
+
+  return std::nullopt;
+}
+```
+:::
+
+
+## Parsing Command-Line Options
+
+Our next example shows how command-line options could be automatically mapped to an "options" structure.
+For simplicity, we posit the existence of a range-like class type `ProgramArgs` that collects the traditional `(argc, argv)` parameters in a more friendly package.
+This example also uses an expansion statement for simplicity.
+
+::: bq
+```c++
+#include <sstream>
+#include <string>
+#include <meta>
+#include <ProgramArgs.h>
+
+template<typename Opts> bool parse_options(Opts *opts, ProgramArgs const &args) {
+  using namespace std;
+  using namespace std::meta;
+  bool success = true;
+  for (auto const arg = args.begin(); args != args.end(); ++args) {
+    template for(constexpr auto dm: members_of(^Opts, is_nonstatic_data_member)) {
```

Comment:
I think it'd be easier to understand the code if we flip the loops:
```cpp
template for (constexpr auto dm : members_of(^Opts, is_nonstatic_data_member)) {
    auto it = ranges::find_if(args, [](string_view arg){ return arg.starts_with("--") && arg.substr(2) == name_of(dm); });
    if (it == args.end()) {
        // not found
    } else if (it + 1 == args.end()) {
        // missing value
    }
    
    // parsing
}
```

Reply:
I'm fine with that.  Can you apply the change you're thinking of?